File: algorithm/code/leetcode/7.reverse-integer/ReverseInteger.java
/**
 * 整数反转
 */
public class ReverseInteger {

    public int run(int x) {
        String s = new StringBuffer(String.valueOf(Math.abs(x))).reverse().toString();
        if (x < 0) {
            s = '-' + s;
        }
        int num;
        try {
            num = Integer.parseInt(s);
        } catch (NumberFormatException e) {
            num = 0;
        }
        return num;
    }

    /**
     * 字符串反转效率比较低
     * 直接对整数进行循环，
     * num = num * 10 + x%10;
     * x /= 10;
     * 注意判断上下越界情况
     *
     * @param x
     * @return
     */
    public int opt(int x) {
        int num = 0;
        while (x != 0) {
            int d = x % 10;
            x = x / 10;
            if (num > Integer.MAX_VALUE / 10 || (Integer.MAX_VALUE / 10 == num && d > 7)) {
                return 0;
            }
            if (num < Integer.MIN_VALUE / 10 || (Integer.MIN_VALUE / 10 == num && d < -8)) {
                return 0;
            }
            num = num * 10 + d;
        }

        return num;
    }
}

File: algorithm/code/leetcode/7.reverse-integer/reverse-integer.go
package main

import "fmt"

// 3121 -> 1213 ((1 * 10 + 2 ) * 10 + 1) * 10 + 3   
func reverse(x int) int {
	maxInt := 1<<31 - 1
	minInt := -1 << 31	
	y := 0
	for x != 0 {
		num := x % 10
		if y > maxInt/10 || (y == maxInt/10 && num > 7) {
			return 0
		}
		if y < minInt/10 || (y == minInt/10 && num < -8) {
			return 0
		}
		y = y * 10 + num
		x = x / 10
	}

	return y
}


func main() {
	
	x := 1534236469
	fmt.Println(x)
	res := reverse(x)
	fmt.Println(res)
}